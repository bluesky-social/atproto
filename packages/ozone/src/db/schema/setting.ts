import { Generated, GeneratedAlways } from 'kysely'
import { Member } from './member'

export const settingTableName = 'setting'

export type SettingScope = 'personal' | 'instance'

export interface Setting {
  id: GeneratedAlways<number>
  key: string
  value: Record<string, unknown>
  managerRole: Member['role'] | null
  description: string | null
  did: string
  scope: SettingScope
  lastUpdatedBy: string
  createdBy: string
  createdAt: Generated<Date>
  updatedAt: Generated<Date>
}

export type PartialDB = {
  [settingTableName]: Setting
}
