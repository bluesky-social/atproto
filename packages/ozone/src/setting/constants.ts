export const ProtectedTagSettingKey = 'tools.ozone.setting.protectedTags'
