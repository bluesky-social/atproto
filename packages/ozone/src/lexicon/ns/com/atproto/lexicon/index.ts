/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  type Auth,
  type Options as XrpcOptions,
  Server as XrpcServer,
  type StreamConfigOrHandler,
  type MethodConfigOrHandler,
  createServer as createXrpcServer,
} from '@atproto/xrpc-server'
import * as ComAtprotoLexiconSchema from '../../../../types/com/atproto/lexicon/schema.js'
import { Server } from '../../../../index.js'

export class ComAtprotoLexiconNS {
  _server: Server

  constructor(server: Server) {
    this._server = server
  }
}
