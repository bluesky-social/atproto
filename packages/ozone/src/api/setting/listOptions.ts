import { AuthRequiredError } from '@atproto/xrpc-server'
import { Server } from '../../lexicon'
import AppContext from '../../context'

export default function (server: Server, ctx: AppContext) {
  server.tools.ozone.setting.listOptions({
    auth: ctx.authVerifier.modOrAdminToken,
    handler: async ({ params, auth }) => {
      const access = auth.credentials
      const db = ctx.db
      const { prefix, scope, keys, limit, cursor } = params
      let did = ctx.cfg.service.did

      if (scope === 'personal') {
        if (access.type !== 'moderator') {
          throw new AuthRequiredError(
            'Must use moderator auth to get personal set details',
          )
        }

        did = access.iss
      }

      const settingService = ctx.settingService(db)

      const result = await settingService.query({
        scope: scope === 'personal' ? 'personal' : 'instance',
        did,
        keys,
        prefix,
        limit,
        cursor,
      })

      return {
        encoding: 'application/json',
        body: {
          options: result.options.map((option) => settingService.view(option)),
          cursor: result.cursor,
        },
      }
    },
  })
}
