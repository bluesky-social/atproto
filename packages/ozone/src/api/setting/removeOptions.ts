import { AuthRequiredError } from '@atproto/xrpc-server'
import { Server } from '../../lexicon'
import AppContext from '../../context'
import { Member } from '../../db/schema/member'

export default function (server: Server, ctx: AppContext) {
  server.tools.ozone.setting.removeOptions({
    auth: ctx.authVerifier.modOrAdminToken,
    handler: async ({ input, auth }) => {
      const access = auth.credentials
      const db = ctx.db
      const { keys, scope } = input.body
      let did = ctx.cfg.service.did
      let managerRole: Member['role'][] = []

      if (scope === 'personal') {
        if (access.type !== 'moderator') {
          throw new AuthRequiredError(
            'Must use moderator auth to delete personal setting',
          )
        }

        did = access.iss
      }

      // When attempting to delete an instance setting using admin_token will allow removing any setting
      // otherwise, admins can remove settings that are manageable by all roles
      // moderators can remove settings that are manageable by moderator and triage roles
      // triage can remove settings that are manageable by triage role
      if (scope === 'instance') {
        managerRole = [
          'tools.ozone.team.defs#roleModerator',
          'tools.ozone.team.defs#roleTriage',
          'tools.ozone.team.defs#roleAdmin',
        ]

        if (access.type !== 'admin_token' && !access.isAdmin) {
          if (access.isModerator) {
            managerRole = [
              'tools.ozone.team.defs#roleModerator',
              'tools.ozone.team.defs#roleTriage',
            ]
          } else if (access.isTriage) {
            managerRole = ['tools.ozone.team.defs#roleTriage']
          }
        }
      }

      const settingService = ctx.settingService(db)

      await settingService.removeOptions(keys, {
        scope: scope === 'personal' ? 'personal' : 'instance',
        managerRole,
        did,
      })

      return {
        encoding: 'application/json',
        body: {},
      }
    },
  })
}
