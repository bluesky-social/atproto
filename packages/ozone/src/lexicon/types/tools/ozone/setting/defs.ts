/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { ValidationResult, BlobRef } from '@atproto/lexicon'
import { CID } from 'multiformats/cid'
import { lexicons } from '../../../../lexicons'
import { $Type, is$typed } from '../../../../util'

const id = 'tools.ozone.setting.defs'

export interface Option {
  key: string
  did: string
  value: {}
  description?: string
  createdAt?: string
  updatedAt?: string
  managerRole?:
    | 'tools.ozone.team.defs#roleModerator'
    | 'tools.ozone.team.defs#roleTriage'
    | 'tools.ozone.team.defs#roleAdmin'
    | (string & {})
  scope: 'instance' | 'personal' | (string & {})
  createdBy: string
  lastUpdatedBy: string
  [k: string]: unknown
}

export function isOption(
  v: unknown,
): v is Option & { $type: $Type<'tools.ozone.setting.defs', 'option'> } {
  return is$typed(v, id, 'option')
}

export function validateOption(v: unknown) {
  return lexicons.validate(`${id}#option`, v) as ValidationResult<Option>
}
