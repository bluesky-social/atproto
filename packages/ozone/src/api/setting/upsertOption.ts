import { AuthRequiredError } from '@atproto/xrpc-server'
import { Server } from '../../lexicon'
import AppContext from '../../context'
import { AdminTokenOutput, ModeratorOutput } from '../../auth-verifier'
import { SettingService } from '../../setting/service'
import { Member } from '../../db/schema/member'
import { ToolsOzoneTeamDefs } from '@atproto/api'
import assert from 'node:assert'
import { settingValidators } from '../../setting/validators'

export default function (server: Server, ctx: AppContext) {
  server.tools.ozone.setting.upsertOption({
    auth: ctx.authVerifier.modOrAdminToken,
    handler: async ({ input, auth }) => {
      const access = auth.credentials
      const db = ctx.db
      const { key, value, description, managerRole, scope } = input.body
      const serviceDid = ctx.cfg.service.did
      let ownerDid = serviceDid

      if (scope === 'personal' && access.type !== 'moderator') {
        throw new AuthRequiredError(
          'Must use moderator auth to create or update a personal setting',
        )
      }

      // if the caller is using moderator auth and storing personal setting
      // use the caller's DID as the owner
      if (scope === 'personal' && access.type === 'moderator') {
        ownerDid = access.iss
      }

      const now = new Date()
      const baseOption = {
        key,
        value,
        did: ownerDid,
        createdBy: ownerDid,
        lastUpdatedBy: ownerDid,
        description: description || '',
        createdAt: now,
        updatedAt: now,
      }

      const settingService = ctx.settingService(db)
      if (scope === 'personal') {
        await settingService.upsert({
          ...baseOption,
          scope: 'personal',
          managerRole: null,
        })
      } else {
        const manageableRoles = getRolesForInstanceOption(access)
        const existingSetting = await getExistingSetting(
          settingService,
          ownerDid,
          key,
          'instance',
        )

        if (
          existingSetting?.managerRole &&
          !manageableRoles.includes(existingSetting.managerRole)
        ) {
          throw new AuthRequiredError(`Not permitted to update setting ${key}`)
        }
        const option = {
          ...baseOption,
          scope: 'instance' as const,
          managerRole: getManagerRole(managerRole),
        }

        if (settingValidators.has(key)) {
          await settingValidators.get(key)?.(option)
        }

        await settingService.upsert(option)
      }

      const newOption = await getExistingSetting(
        settingService,
        ownerDid,
        key,
        scope,
      )
      assert(newOption, 'Failed to get the updated setting')

      return {
        encoding: 'application/json',
        body: {
          option: settingService.view(newOption),
        },
      }
    },
  })
}

const getExistingSetting = async (
  settingService: SettingService,
  did: string,
  key: string,
  scope: string,
) => {
  const result = await settingService.query({
    scope: scope === 'personal' ? 'personal' : 'instance',
    keys: [key],
    limit: 1,
    did,
  })

  return result.options[0]
}

const getRolesForInstanceOption = (
  access: AdminTokenOutput['credentials'] | ModeratorOutput['credentials'],
) => {
  const fullPermission = [
    ToolsOzoneTeamDefs.ROLEADMIN,
    ToolsOzoneTeamDefs.ROLEMODERATOR,
    ToolsOzoneTeamDefs.ROLETRIAGE,
  ]
  if (access.type === 'admin_token') {
    return fullPermission
  }

  if (access.isAdmin) {
    return fullPermission
  }

  if (access.isModerator) {
    return [ToolsOzoneTeamDefs.ROLEMODERATOR, ToolsOzoneTeamDefs.ROLETRIAGE]
  }

  return [ToolsOzoneTeamDefs.ROLETRIAGE]
}

const getManagerRole = (role?: string) => {
  let managerRole: Member['role'] | null = null

  if (role === ToolsOzoneTeamDefs.ROLEADMIN) {
    managerRole = ToolsOzoneTeamDefs.ROLEADMIN
  } else if (role === ToolsOzoneTeamDefs.ROLEMODERATOR) {
    managerRole = ToolsOzoneTeamDefs.ROLEMODERATOR
  } else if (role === ToolsOzoneTeamDefs.ROLETRIAGE) {
    managerRole = ToolsOzoneTeamDefs.ROLETRIAGE
  }

  return managerRole
}
