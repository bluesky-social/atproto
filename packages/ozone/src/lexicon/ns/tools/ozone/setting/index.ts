/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  type Auth,
  type Options as XrpcOptions,
  Server as XrpcServer,
  type StreamConfigOrHandler,
  type MethodConfigOrHandler,
  createServer as createXrpcServer,
} from '@atproto/xrpc-server'
import * as ToolsOzoneSettingListOptions from '../../../../types/tools/ozone/setting/listOptions.js'
import * as ToolsOzoneSettingRemoveOptions from '../../../../types/tools/ozone/setting/removeOptions.js'
import * as ToolsOzoneSettingUpsertOption from '../../../../types/tools/ozone/setting/upsertOption.js'
import { Server } from '../../../../index.js'

export class ToolsOzoneSettingNS {
  _server: Server

  constructor(server: Server) {
    this._server = server
  }

  listOptions<A extends Auth = void>(
    cfg: MethodConfigOrHandler<
      A,
      ToolsOzoneSettingListOptions.QueryParams,
      ToolsOzoneSettingListOptions.HandlerInput,
      ToolsOzoneSettingListOptions.HandlerOutput
    >,
  ) {
    const nsid = 'tools.ozone.setting.listOptions' // @ts-ignore
    return this._server.xrpc.method(nsid, cfg)
  }

  removeOptions<A extends Auth = void>(
    cfg: MethodConfigOrHandler<
      A,
      ToolsOzoneSettingRemoveOptions.QueryParams,
      ToolsOzoneSettingRemoveOptions.HandlerInput,
      ToolsOzoneSettingRemoveOptions.HandlerOutput
    >,
  ) {
    const nsid = 'tools.ozone.setting.removeOptions' // @ts-ignore
    return this._server.xrpc.method(nsid, cfg)
  }

  upsertOption<A extends Auth = void>(
    cfg: MethodConfigOrHandler<
      A,
      ToolsOzoneSettingUpsertOption.QueryParams,
      ToolsOzoneSettingUpsertOption.HandlerInput,
      ToolsOzoneSettingUpsertOption.HandlerOutput
    >,
  ) {
    const nsid = 'tools.ozone.setting.upsertOption' // @ts-ignore
    return this._server.xrpc.method(nsid, cfg)
  }
}
