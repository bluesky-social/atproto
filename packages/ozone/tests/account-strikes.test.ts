import AtpAgent from '@atproto/api'
import {
  ModeratorClient,
  SeedClient,
  TestNetwork,
  basicSeed,
} from '@atproto/dev-env'
import { ids } from '../src/lexicon/lexicons'
import { SeverityLevelSettingKey } from '../src/setting/constants'
import { forSnapshot } from './_util'

const strikeConfig = {
  'sev-0': { strikeCount: 0 },
  'sev-1': {
    strikeCount: 1,
    strikeOnOccurrence: 2,
    expiresInDays: 365,
  },
  'sev-2': { strikeCount: 2, expiresInDays: 365 },
  'sev-4': { strikeCount: 4, expiresInDays: 0 },
  'sev-5': { needsTakedown: true },
}

describe('account-strikes', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let sc: SeedClient
  let modClient: ModeratorClient

  const repoSubject = (did: string) => ({
    $type: 'com.atproto.admin.defs#repoRef',
    did,
  })

  const configureSeverityLevels = async () => {
    // Configure severity level settings
    await agent.tools.ozone.setting.upsertOption(
      {
        scope: 'instance',
        key: SeverityLevelSettingKey,
        value: strikeConfig,
        description: 'Severity level configuration for strike system',
        managerRole: 'tools.ozone.team.defs#roleAdmin',
      },
      {
        encoding: 'application/json',
        headers: await network.ozone.modHeaders(
          ids.ToolsOzoneSettingUpsertOption,
          'admin',
        ),
      },
    )
  }

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_account_strikes',
    })
    agent = network.ozone.getClient()
    sc = network.getSeedClient()
    modClient = network.ozone.getModClient()
    await basicSeed(sc)
    await network.processAll()
    await configureSeverityLevels()
  })

  afterAll(async () => {
    await network.close()
  })

  it('tracks strikes and exposes them through queryStatuses and queryEvents', async () => {
    const aliceSubject = repoSubject(sc.dids.alice)
    const alicePost = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.alice][0].ref.uriStr,
      cid: sc.posts[sc.dids.alice][0].ref.cidStr,
    }

    await modClient.emitEvent({
      event: {
        $type: 'tools.ozone.moderation.defs#modEventTakedown',
        severityLevel: 'sev-2',
        strikeCount: strikeConfig['sev-2'].strikeCount,
        comment: 'First violation',
      },
      subject: alicePost,
    })

    const alicePost2 = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.alice][1].ref.uriStr,
      cid: sc.posts[sc.dids.alice][1].ref.cidStr,
    }
    await modClient.emitEvent({
      event: {
        $type: 'tools.ozone.moderation.defs#modEventTakedown',
        severityLevel: 'sev-2',
        strikeCount: strikeConfig['sev-2'].strikeCount,
        comment: 'Second violation',
      },
      subject: alicePost2,
    })

    const alicePost3 = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.alice][2].ref.uriStr,
      cid: sc.posts[sc.dids.alice][2].ref.cidStr,
    }
    await modClient.emitEvent({
      event: {
        $type: 'tools.ozone.moderation.defs#modEventTakedown',
        severityLevel: 'sev-1',
        strikeCount: 0, // First occurrence - warning
        policies: ['spam-policy'],
        comment: 'First sev-1 violation',
      },
      subject: alicePost3,
    })

    // Issue second sev-1 takedown with same policy on another post (second occurrence, should be 1 strike, total 5)
    const aliceReply = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.replies[sc.dids.alice][0].ref.uriStr,
      cid: sc.replies[sc.dids.alice][0].ref.cidStr,
    }
    await modClient.emitEvent({
      event: {
        $type: 'tools.ozone.moderation.defs#modEventTakedown',
        severityLevel: 'sev-1',
        strikeCount: strikeConfig['sev-1'].strikeCount, // Second occurrence - actual strike
        policies: ['spam-policy'],
        comment: 'Second sev-1 violation',
      },
      subject: aliceReply,
    })

    await modClient.emitEvent({
      event: {
        $type: 'tools.ozone.moderation.defs#modEventTakedown',
        severityLevel: 'sev-0',
        strikeCount: strikeConfig['sev-0'].strikeCount,
        comment: 'Warning only',
      },
      subject: aliceSubject,
    })

    const statusResult = await modClient.queryStatuses({
      subject: sc.dids.alice,
    })

    expect(statusResult.subjectStatuses.length).toBeGreaterThan(0)
    expect(forSnapshot(statusResult.subjectStatuses[0])).toMatchSnapshot()

    // Filter for events with at least 2 strikes
    const eventsWithMinStrikes = await modClient.queryEvents({
      subject: sc.dids.alice,
      includeAllUserRecords: true,
      types: ['tools.ozone.moderation.defs#modEventTakedown'],
      minStrikeCount: 2,
    })
    expect(forSnapshot(eventsWithMinStrikes.events)).toMatchSnapshot()
  })
})
