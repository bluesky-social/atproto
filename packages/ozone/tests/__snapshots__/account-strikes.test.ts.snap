// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`account-strikes tracks strikes and exposes them through queryStatuses and queryEvents 1`] = `
Object {
  "accountStats": Object {
    "$type": "tools.ozone.moderation.defs#accountStats",
  },
  "accountStrike": Object {
    "$type": "tools.ozone.moderation.defs#accountStrike",
    "activeStrikeCount": 5,
    "firstStrikeAt": "1970-01-01T00:00:00.000Z",
    "lastStrikeAt": "1970-01-01T00:00:00.000Z",
    "totalStrikeCount": 5,
  },
  "ageAssuranceState": "unknown",
  "createdAt": "1970-01-01T00:00:00.000Z",
  "hosting": Object {
    "$type": "tools.ozone.moderation.defs#accountHosting",
    "status": "unknown",
  },
  "id": 9,
  "lastReviewedAt": "1970-01-01T00:00:00.000Z",
  "lastReviewedBy": "user(0)",
  "priorityScore": 0,
  "recordsStats": Object {
    "$type": "tools.ozone.moderation.defs#recordsStats",
  },
  "reviewState": "tools.ozone.moderation.defs#reviewClosed",
  "subject": Object {
    "$type": "com.atproto.admin.defs#repoRef",
    "did": "user(1)",
  },
  "subjectBlobCids": Array [],
  "subjectRepoHandle": "alice.test",
  "tags": Array [
    "lang:und",
  ],
  "takendown": true,
  "updatedAt": "1970-01-01T00:00:00.000Z",
}
`;

exports[`account-strikes tracks strikes and exposes them through queryStatuses and queryEvents 2`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(0)",
    "event": Object {
      "$type": "tools.ozone.moderation.defs#modEventTakedown",
      "comment": "Second violation",
      "severityLevel": "sev-2",
      "strikeCount": 2,
    },
    "id": 3,
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(0)",
      "uri": "record(0)",
    },
    "subjectBlobCids": Array [],
    "subjectHandle": "alice.test",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(0)",
    "event": Object {
      "$type": "tools.ozone.moderation.defs#modEventTakedown",
      "comment": "First violation",
      "severityLevel": "sev-2",
      "strikeCount": 2,
    },
    "id": 1,
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(1)",
      "uri": "record(1)",
    },
    "subjectBlobCids": Array [],
    "subjectHandle": "alice.test",
  },
]
`;
